infer/internal/ende/ende.go

```diff
@@ -338,7 +361,38 @@ const (
 	isEmptyArr = iota
 )
 
-// Mark a encoder as generating values only.
+func flattenAssets(a any) (resource.PropertyValue, bool) {
```

Comment:
Here's an example of a function we could extract to make this handling easier:

```go
func extractAssetOrArchive(aMap map[string]any, typeSig, kindSig string) (map[string]any, bool) {
	insideType, hasTypeSig := aMap[kindSig]
	if !hasTypeSig {
		return nil, false
	}
	insideTypeMap, ok := insideType.(map[string]any)
	if !ok {
		return nil, false
	}
	kind, hasKind := insideTypeMap[sig.Key]
	if !hasKind {
		return nil, false
	}
	kindStr, ok := kind.(string)
	if !ok {
		return nil, false
	}
	if kindStr != typeSig {
		return nil, false
	}
	return insideTypeMap, true
}
```

We can then be more specific on how we want to handle each error case too - panic or error etc.

Reply:
The method above then simplifies to:

```go
	if aMap, ok := a.(map[string]any); ok {
		if asAsset, isAsset := extractAssetOrArchive(aMap, sig.AssetSig, types.AssetSignature); isAsset {
			return resource.NewObjectProperty(resource.NewPropertyMapFromMap(asAsset)), true
		}
		if asArchive, isArchive := extractAssetOrArchive(aMap, sig.ArchiveSig, types.ArchiveSignature); isArchive {
			return resource.NewObjectProperty(resource.NewPropertyMapFromMap(asArchive)), true
		}
	}
````

Reply:
I don't quite see the simplification here TBH. I've commented and slightly reorganized the method in the latest commit, take a look if that looks better to you.

Reply:
It's certainly partially subjective, so I wouldn't block on this.

Some observations of what makes it look complex to me are:
1. 5 levels of nested if statements means that the error handling paths are somewhat implicit rather than explicit - falling through to the null return (unless it ends up at the panic). The `extractAssetOrArchive` style would give the option for returning specific error messages for each case if we needed them too easier than adding `else` blocks for each `if` level. 
2. The panic "Encountered an unknown kind in an AssetOrArchive" is not adjacent to the condition that would trigger it - making the reader track back to work out the context. I.e. it's inverse of the branch it didn't take 4 lines above (`kind == sig.AssetSig || kind == sig.ArchiveSig`).
3. It's tricky follow if we're looking at an asset or we're looking at an archive, though it then appears it kinda doesn't matter after line 765 when we set the `raw` variable.

Reply:
The paths leading to the default return are not error paths. It's entirely expected that we encounter maps that are not AssetOrArchive, and fall through.

Reply:
There's also implicit error paths too:
- If it's got the asset type sig, but it's not a map within
- If it's an asset type sig, but has no kind field
- The kind field is not a string

These all just return null, which I think it fine, but it's implicit.